src/core/atomic.d

```diff
@@ -109,7 +109,7 @@ enum MemoryOrder
  * Returns:
  *  The value of 'val'.
  */
-TailShared!T atomicLoad(MemoryOrder ms = MemoryOrder.seq, T)( ref const shared T val ) pure nothrow @nogc @trusted
+inout(TailShared!T) atomicLoad(MemoryOrder ms = MemoryOrder.seq, T)( ref inout shared T val ) pure nothrow @nogc @trusted
```

Comment:
As far as I can tell, this `inout` stuff isn't necessary.

Going from `const T` to `T` is fine if it's a copy. Only the (copied) head of the type is made mutable.

Example:

```d
import core.atomic;
shared const int* x;
pragma(msg, typeof(atomicLoad(x))); /* shared(const(int))* */
```

Note that the referenced `int` is still `const`.

Reply:
Is like to think that's true, I didn't write this and it's a huge mess, but if I change anything at all it all breaks spectacularly.

I think some of the downstream code is maladjusted to this complexity, so it would be good to tidy up generally, but that's a different PR.

Reply:
The worst part here is that atomics shouldn't be shared at all... unless the user passed a shared thing, in which case it should work properly with that.

Reply:
> I didn't write this and it's a huge mess, but if I change anything at all it all breaks spectacularly.

You added `inout` which is what I'm commenting on. I'm saying you should be able to leave it as it is (`const` instead of `inout`).

I expect that the DMD bug you filed is the real issue, and I hope that we can fix it.

Reply:
The `T` in the code should contain and propagate the qualifiers, I'd like to get the explicit qualifiers out of the code.

Reply:
So you want to get mutable results when possible? Makes sense. I thought you were getting a mutable result where you don't want one. Never mind then.

Reply:
This should just return a copy of T, whatever T is, and comply with whether T is copyable or not.
Theres this whole `TailConst` machine in `core.atomic`; I have no idea what it does, or why it's there, but it's very complicated and seems to make problems.

Reply:
I'm the author of `TailShared`. It's supposed to be an improvement over the previous `HeadUnshared` which would strip `shared` to eagerly (#1605).

I wouldn't mind if you got rid of it, changing the return type to simply `T`. But changing return types is always risky business, and `shared` is still underspecified, as far as I know. So that's probably not going to be easy.

Reply:
Oh perfect!

Well, I need to understand what it does and why it's there before removing it... changing API in here breaks lots of client code, so I haven't been able to make much simplification.
But I also think some of the unittests have bugs which are there to hide some bugs in the implementation... like a double-negative situation.

I had trouble with it because it wraps the type and changes the calling convention. It generates extra code in loadAtomic by doing a copy from one calling convention to another.
Slice/delegate returns in RAX;RDX, but TailShared returns by stack, so you can see `atomicLoad` has that thunk in the return statement which is there so that it transfers the return value from one ABI to the other. It would be better if it wasn't there... but like I say, I don't understand the problem it was trying to solve.

I've been working with Andrei and Walter and friends on what shared needs to be in the future... we should talk, do you have IM?
Have you been contributing to `core.atomic` a lot? (Your github profile gives me no idea who you are)

Reply:
> Well, I need to understand what it does and why it's there before removing it... changing API in here breaks lots of client code, so I haven't been able to make much simplification.

`TailShared` does (or is supposed to do) what it says in the comment: "Construct a type with a shared tail, and if possible with an unshared head."

`shared(T*)` becomes `shared(T)*`. As for why that's done, I suppose it's because dealing with `shared` is cumbersome, so it's removed as much as possible. But I didn't add that, I just refined it. I can't say for sure what the original motivation was.

`T*` also becomes `shared(T)*`. If I remember correctly, this was just easier to implement and/or use, because `atomicLoad` and friends always operate on `shared` types anyway.

`shared SomeClass` remains `shared SomeClass`. This is the whole improvement over `HeadUnshared` which would remove `shared`.

> But I also think some of the unittests have bugs which are there to hide some bugs in the implementation... like a double-negative situation.

Which tests do you think are wrong?

> I had trouble with it because it wraps the type and changes the calling convention. It generates extra code in loadAtomic by doing a copy from one calling convention to another.
> Slice/delegate returns in RAX;RDX, but TailShared returns by stack, so you can see `atomicLoad` has that thunk in the return statement which is there so that it transfers the return value from one ABI to the other. It would be better if it wasn't there... but like I say, I don't understand the problem it was trying to solve.

Being a type template, `TailShared` doesn't return anything in the sense of calling conventions. And (unfortunately) I can't shed light on what `atomicLoad` does. I didn't write it and I know next to nothing about atomic operations, memory fences, etc. Sorry.

> I've been working with Andrei and Walter and friends on what shared needs to be in the future... we should talk, do you have IM?

You can email me at gmail, but I don't have anything to offer when it comes to technical details of dealing with shared data. At best I can make amateurish comments on the type system side of things.

> Have you been contributing to `core.atomic` a lot?

No. I'm pretty sure that turning `HeadUnshared` into `TailShared` was my most significant contribution here. Might actually be the only one. And that required no actual knowledge of atomics, just of D's type system.

Sorry again if I got your hopes up. I'm not actually knowledgeable about most of this.

Reply:
> > I had trouble with it because it wraps the type and changes the calling convention. It generates extra code in loadAtomic by doing a copy from one calling convention to another.
> > Slice/delegate returns in RAX;RDX, but TailShared returns by stack, so you can see `atomicLoad` has that thunk in the return statement which is there so that it transfers the return value from one ABI to the other. It would be better if it wasn't there... but like I say, I don't understand the problem it was trying to solve.
> 
> Being a type template, `TailShared` doesn't return anything in the sense of calling conventions.

Well it does, because the function `T fun()`, and `TailShared!T fun()` have a different calling convention in some cases... which is problematic when there's thunk's all over the code. I'm trying to remove all the ugly shit as best I can.

I've made some changes that isolate `TailShared` to a single overload which operates on that exact case, and tried to narrow it down.
I think it's better now, and the tests pass, but what I think might be better if the user called the function with a `TailShared!T` manually. They can use it when they want it.